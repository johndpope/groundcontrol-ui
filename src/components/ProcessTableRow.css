.ProcessTableRow td {
  font: var(--mono-font);
}

.ProcessTableRowCommand {
  word-break: break-all;
  white-space: pre-wrap;
}

.ui.table td.ProcessTableRowActions {
  padding-top: 0;
  padding-bottom: 0;
}
