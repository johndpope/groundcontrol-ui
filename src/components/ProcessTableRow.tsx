// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import graphql from "babel-plugin-relay/macro";
import React from "react";
import { createFragmentContainer } from "react-relay";
import {
  Button,
  Table,
 } from "semantic-ui-react";

import { ProcessTableRow_item } from "./__generated__/ProcessTableRow_item.graphql";

import RepositoryShortName from "./RepositoryShortName";

import "./ProcessTableRow.css";

export interface IProps {
  item: ProcessTableRow_item;
  onStartProcess: (values: IProps) => any;
  onStopProcess: (values: IProps) => any;
}

export function ProcessTableRow(props: IProps) {
  const {
    item: {
      command,
      status,
      project: {
        repository,
      },
    },
    onStartProcess,
    onStopProcess,
  } = props;
  const buttons: JSX.Element[] = [];
  const handleStart = () => onStartProcess(props);
  const handleStop = () => onStopProcess(props);

  switch (status) {
  case "DONE":
  case "FAILED":
    buttons.push((
      <Button
        key="start"
        compact={true}
        size="mini"
        icon="play"
        content="Start"
        onClick={handleStart}
      />
    ));
    break;
  case "RUNNING":
    buttons.push((
      <Button
        key="stop"
        size="mini"
        compact={true}
        icon="stop"
        content="Stop"
        onClick={handleStop}
      />
    ));
    break;
  case "STOPPING":
    buttons.push((
      <Button
        key="stop"
        size="mini"
        compact={true}
        icon="stop"
        content="Stop"
        disabled={true}
        loading={true}
      />
    ));
    break;
  }

  return (
    <Table.Row className="ProcessTableRow">
      <Table.Cell collapsing={true}>
        <RepositoryShortName repository={repository} />
      </Table.Cell>
      <Table.Cell
        className="ProcessTableRowCommand"
      >
        {command}
      </Table.Cell>
      <Table.Cell
        positive={status === "DONE"}
        warning={status === "RUNNING"}
        error={status === "FAILED"}
        collapsing={true}
      >
        {status}
      </Table.Cell>
      <Table.Cell
        className="ProcessTableRowActions"
        collapsing={true}
      >
        {buttons}
      </Table.Cell>
    </Table.Row>
  );
}

export default createFragmentContainer(ProcessTableRow, graphql`
  fragment ProcessTableRow_item on Process {
    id
    command
    status
    project {
      repository
    }
  }`,
);
