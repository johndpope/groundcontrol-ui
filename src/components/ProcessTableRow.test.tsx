// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { shallow } from "enzyme";
import React from "react";
import { mocked } from "ts-jest/utils";

import { ProcessTableRow_item } from "./__generated__/ProcessTableRow_item.graphql";

import { mockQueryPropAttrs } from "../testing/relay";

import { IProps, ProcessTableRow } from "./ProcessTableRow";

jest.mock("./RepositoryShortName", () => "RepositoryShortName");

const item: ProcessTableRow_item = {
  ...mockQueryPropAttrs(),
  command: "command",
  id: "id",
  project: {
    repository: "repository",
  },
  status: "RUNNING",
};

const props = {
  item,
  onStartProcess: jest.fn(),
  onStopProcess: jest.fn(),
};

beforeEach(() => {
  mocked(props.onStartProcess).mockClear();
  mocked(props.onStopProcess).mockClear();
});

describe("<ProcessTableRow />", () => {

  it("renders correctly", () => {
    const wrapper = shallow(<ProcessTableRow {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("shows a stop button if the process is running", () => {
    const wrapper = shallow(
      <ProcessTableRow
        {...{...props, item: { ...item, status: "RUNNING" } }}
      />,
    );
    expect(wrapper.find("Button[icon='stop']")).toHaveLength(1);
  });

  it("shows a start button if the process is done", () => {
    const wrapper = shallow(
      <ProcessTableRow
        {...{...props, item: { ...item, status: "DONE" } }}
      />,
    );
    expect(wrapper.find("Button[icon='play']")).toHaveLength(1);
  });

  it("shows a start button if the process failed", () => {
    const wrapper = shallow(
      <ProcessTableRow
        {...{...props, item: { ...item, status: "FAILED" } }}
      />,
    );
    expect(wrapper.find("Button[icon='play']")).toHaveLength(1);
  });

  it("shows a disabled stop button if the process is stopping", () => {
    const wrapper = shallow(
      <ProcessTableRow
        {...{...props, item: { ...item, status: "STOPPING" } }}
      />,
    );
    expect(wrapper.find("Button[icon='stop'][disabled=true]")).toHaveLength(1);
  });

  it("triggers onStopProcess when the stop button is clicked", () => {
    const wrapper = shallow(
      <ProcessTableRow
        {...{...props, item: { ...item, status: "RUNNING" } }}
      />,
    );
    wrapper.find("Button[icon='stop']").simulate("click");
    expect(props.onStopProcess).toBeCalledTimes(1);
    expect(props.onStopProcess).toBeCalledWith(props);
  });

  it("triggers onStartProcess when the process button is clicked", () => {
    const doneProps: IProps = {...props, item: { ...item, status: "DONE" } };
    const wrapper = shallow(<ProcessTableRow {...doneProps} />);
    wrapper.find("Button[icon='play']").simulate("click");
    expect(doneProps.onStartProcess).toBeCalledTimes(1);
    expect(doneProps.onStartProcess).toBeCalledWith(doneProps);
  });

});
