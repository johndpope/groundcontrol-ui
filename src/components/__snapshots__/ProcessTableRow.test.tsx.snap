// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProcessTableRow /> renders correctly 1`] = `
<TableRow
  as="tr"
  cellAs="td"
  className="ProcessTableRow"
>
  <TableCell
    as="td"
    collapsing={true}
  >
    <RepositoryShortName
      repository="repository"
    />
  </TableCell>
  <TableCell
    as="td"
    className="ProcessTableRowCommand"
  >
    command
  </TableCell>
  <TableCell
    as="td"
    collapsing={true}
    error={false}
    positive={false}
    warning={true}
  >
    RUNNING
  </TableCell>
  <TableCell
    as="td"
    className="ProcessTableRowActions"
    collapsing={true}
  >
    <Button
      as="button"
      compact={true}
      content="Stop"
      icon="stop"
      key="stop"
      onClick={[Function]}
      size="mini"
    />
  </TableCell>
</TableRow>
`;
